File: mdl2obj.c
/*
 * mdl2obj
 *
 * Copyright (C) 2016 Florian Zwoch <fzwoch@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

#include <glib.h>
#include <string.h>

struct vec3_s {
	gfloat x;
	gfloat y;
	gfloat z;
};

struct mdl_s {
	guint32 id;
	guint32 version;
	struct vec3_s scale;
	struct vec3_s origin;
	gfloat radius;
	struct vec3_s offsets;
	guint32 num_skins;
	guint32 skin_width;
	guint32 skin_height;
	guint32 num_verts;
	guint32 num_triangles;
	guint32 num_frames;
	guint32 sync_type;
	guint32 flags;
	gfloat size;
};

struct skin_s {
	guint32 type;
};

struct skin_group_s {
	guint32 type;
	guint32 num_skins;
	gfloat time;
};

struct stvert_s {
	guint32 on_seam;
	guint32 s;
	guint32 t;
};

struct triangle_s {
	guint32 front;
	guint32 vertex[3];
};

struct vert_s {
	guint8 v[3];
	guint8 normal;
};

struct frame_s {
	guint32 type;
	struct vert_s min;
	struct vert_s max;
	gchar name[16];
};

int main(int argc, char **argv)
{
	GError *err = NULL;
	gchar *buf = NULL;
	gsize buf_len = 0;
	
	if (argc != 2)
	{
		g_print("usage: %s <model.mdl>\n", argv[0]);
		return 1;
	}
	
	g_file_get_contents(argv[1], &buf, &buf_len, &err);
	if (err != NULL)
	{
		g_error("%s", err->message);
		g_error_free(err);
		
		return 1;
	}
	
	struct mdl_s *mdl = buf;
	struct skin_s *skin = mdl + 1;
	
	for (gint i = 0; i < mdl->num_skins; i++)
	{
		if (skin->type != 0)
		{
			struct skin_group_s *skin_group = skin;
			
			skin = (void*)skin + sizeof(struct skin_group_s) + skin_group->num_skins * mdl->skin_width * mdl->skin_height;
		}
		else
		{
			skin = (void*)skin + sizeof(struct skin_s) + mdl->skin_width * mdl->skin_height;
		}
	}
	
	struct stvert_s *stvert = skin;
	struct triangle_s *triangle = stvert + mdl->num_verts;
	struct frame_s *frame = triangle + mdl->num_triangles;
	
	GString *obj = g_string_new(NULL);
	
	gchar *mdl_name = g_path_get_basename(argv[1]);
	mdl_name[strcspn(mdl_name, ".")] = '\0';
	
	g_string_append_printf(obj, "newmtl %s\n", mdl_name);
	g_string_append_printf(obj, "Ka 1 1 1\n");
	g_string_append_printf(obj, "Kd 1 1 1\n");
	g_string_append_printf(obj, "Ks 0 0 0\n");
	g_string_append_printf(obj, "Tr 1\n");
	g_string_append_printf(obj, "illum 1\n");
	g_string_append_printf(obj, "Ns 0\n");
	g_string_append_printf(obj, "map_Kd textures/%s.jpg\n", mdl_name);
	
	gchar *out_file = g_strdup_printf("%s.mtl", mdl_name);
	
	g_file_set_contents(out_file, obj->str, obj->len, &err);
	if (err != NULL)
	{
		g_error("%s", err->message);
		g_error_free(err);
		
		return 1;
	}
	
	for (gint k = 0; k < mdl->num_frames; k++)
	{
		struct vert_s *vert = frame + 1;
		
		if (frame->type != 0)
		{
			frame = (void*)frame + 2 * sizeof(struct vert_s) + sizeof(gfloat);
		}
		
		frame = (void*)frame + sizeof(struct frame_s) + mdl->num_verts * sizeof(struct vert_s);
		
		g_string_printf(obj, "mtllib %s.mtl\n", mdl_name);
		g_string_append_printf(obj, "usemtl %s\n", mdl_name);
		
		for (gint i = 0; i < mdl->num_verts; i++)
		{
			g_string_append_printf(obj, "v %g %g %g\n",
								   mdl->scale.x * vert[i].v[0] + mdl->origin.x,
								   mdl->scale.y * vert[i].v[1] + mdl->origin.y,
								   mdl->scale.z * vert[i].v[2] + mdl->origin.z);
		}

		gboolean needs_hack = FALSE;

		for (gint i = 0; i < mdl->num_verts; i++)
		{
			g_string_append_printf(obj, "vt %g %g\n",
								   (float) stvert[i].s / mdl->skin_width,
								   1 - (float) stvert[i].t / mdl->skin_height);

			if (stvert[i].on_seam != 0)
			{
				needs_hack = TRUE;
			}
		}

		if (needs_hack == TRUE)
		{
			for (gint i = 0; i < mdl->num_verts; i++)
			{
				if (stvert[i].on_seam == 0)
				{
					g_string_append_printf(obj, "vt 0 0\n"); // can never be used
				}
				else
				{
					g_string_append_printf(obj, "vt %g %g\n",
										(float) (stvert[i].s + mdl->skin_width / 2) / mdl->skin_width,
										1 - (float) stvert[i].t / mdl->skin_height);
				}
			}
		}
		
		for (gint i = 0; i < mdl->num_triangles; i++)
		{
			g_string_append_printf(obj, "f %d/%d %d/%d %d/%d\n",
								   triangle[i].vertex[0] + 1,
								   triangle[i].vertex[0] + 1 + (triangle[i].front == 0 && stvert[triangle[i].vertex[0]].on_seam != 0 ? mdl->num_verts : 0),
								   triangle[i].vertex[2] + 1,
								   triangle[i].vertex[2] + 1 + (triangle[i].front == 0 && stvert[triangle[i].vertex[2]].on_seam != 0 ? mdl->num_verts : 0),
								   triangle[i].vertex[1] + 1,
								   triangle[i].vertex[1] + 1 + (triangle[i].front == 0 && stvert[triangle[i].vertex[1]].on_seam != 0 ? mdl->num_verts : 0));
		}
		
		if (mdl->num_frames == 1)
		{
			out_file = g_strdup_printf("%s.obj", mdl_name);
		}
		else
		{
			out_file = g_strdup_printf("%s_%d.obj", mdl_name, k);
		}
		
		g_file_set_contents(out_file, obj->str, obj->len, &err);
		if (err != NULL)
		{
			g_error("%s", err->message);
			g_error_free(err);
			
			return 1;
		}
		
		g_free(out_file);
	}
	
	g_string_free(obj, TRUE);
	g_free(buf);
	
	return 0;
}

File: mdl2obj.go
/*
 * mdl2obj
 *
 * Copyright (C) 2016-2019 Florian Zwoch <fzwoch@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package main

import (
	"bufio"
	"encoding/binary"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strings"
	"text/template"
)

type vec3 struct {
	X float32
	Y float32
	Z float32
}

type mdlHeader struct {
	ID           [4]byte
	Version      uint32
	Scale        vec3
	Origin       vec3
	Radius       float32
	Offsets      vec3
	NumSkins     uint32
	SkinWidth    uint32
	SkinHeight   uint32
	NumVerts     uint32
	NumTriangles uint32
	NumFrames    uint32
	SyncType     uint32
	Flags        uint32
	Size         float32
}

type skin struct {
	Type uint32
}

type skinGroup struct {
	NumSkins uint32
	Time     float32
}

type stVert struct {
	OnSeam uint32
	S      uint32
	T      uint32
}

type triangle struct {
	Front  uint32
	Vertex [3]uint32
}

type vert struct {
	V      [3]uint8
	Normal uint8
}

type frame struct {
	Type uint32
	Min  vert
	Max  vert
	Name [16]uint8
}

func main() {
	if len(os.Args) != 2 {
		fmt.Println("usage:", os.Args[0], "<input.mdl>")
		os.Exit(1)
	}

	f, err := os.Open(os.Args[1])
	if err != nil {
		panic(err)
	}
	defer f.Close()

	mdlName := filepath.Base(strings.TrimSuffix(os.Args[1], filepath.Ext(os.Args[1])))

	var mdl mdlHeader

	err = binary.Read(f, binary.LittleEndian, &mdl)
	if err != nil {
		panic(err)
	}

	if string(mdl.ID[:]) != "IDPO" {
		panic("MDL magic " + string(mdl.ID[:]) + " != IDPO")
	}

	if mdl.Version != 6 {
		panic("MDL version " + string(mdl.Version) + " != 6")
	}

	for i := 0; i < int(mdl.NumSkins); i++ {
		var skin skin

		binary.Read(f, binary.LittleEndian, &skin)

		skinGroup := skinGroup{
			NumSkins: 1,
		}

		if skin.Type != 0 {
			binary.Read(f, binary.LittleEndian, &skinGroup)
		}

		f.Seek(int64(skinGroup.NumSkins*mdl.SkinWidth*mdl.SkinHeight), io.SeekCurrent)
	}

	stverts := make([]stVert, mdl.NumVerts)
	triangles := make([]triangle, mdl.NumTriangles)

	binary.Read(f, binary.LittleEndian, &stverts)
	binary.Read(f, binary.LittleEndian, &triangles)

	verts := [][]vert{
		make([]vert, mdl.NumFrames),
	}

	for i := range verts {
		var frame frame

		binary.Read(f, binary.LittleEndian, &frame)

		if frame.Type != 0 {
			panic("FIXME")
		}

                fmt.Println(frame.Name)

		verts[i] = make([]vert, mdl.NumVerts)
		binary.Read(f, binary.LittleEndian, &verts[i])
	}

	data := struct {
		Name string
		V    [][3]float32
		VT   [][2]float32
		F    [][9]uint32
	}{
		Name: mdlName,
		V:    make([][3]float32, len(verts[0])),
		VT:   make([][2]float32, len(stverts)*2),
		F:    make([][9]uint32, len(triangles)),
	}

	for i, v := range verts[0] {
		data.V[i] = [3]float32{
			-mdl.Scale.X*float32(v.V[0]) - mdl.Origin.X,
			mdl.Scale.Z*float32(v.V[2]) + mdl.Origin.Z,
			mdl.Scale.Y*float32(v.V[1]) + mdl.Origin.Y,
		}
	}

	for i, st := range stverts {
		data.VT[i] = [2]float32{
			float32(st.S) / float32(mdl.SkinWidth),
			float32(1 - float32(st.T)/float32(mdl.SkinHeight)),
		}
		data.VT[i+int(mdl.NumVerts)] = [2]float32{
			(float32(st.S) + float32(mdl.SkinWidth/2)) / float32(mdl.SkinWidth),
			1 - float32(st.T)/float32(mdl.SkinHeight),
		}
	}

	for i, t := range triangles {
		var uv [3]uint32

		for i, v := range t.Vertex {
			uv[i] = v
			if t.Front == 0 && stverts[v].OnSeam != 0 {
				uv[i] += mdl.NumVerts
			}
		}

		data.F[i] = [9]uint32{
			t.Vertex[0] + 1,
			uv[0] + 1,
                        uint32(verts[0][t.Vertex[0]].Normal),
			t.Vertex[2] + 1,
			uv[2] + 1,
                        uint32(verts[0][t.Vertex[1]].Normal),
			t.Vertex[1] + 1,
			uv[1] + 1,
                        uint32(verts[0][t.Vertex[2]].Normal),
		}
	}

	o, err := os.Create(mdlName + ".obj")
	if err != nil {
		panic(err)
	}
	w := bufio.NewWriter(o)

	t, _ := template.New("").Parse(obj)
	t.Execute(w, data)
	w.Flush()
	o.Close()

	o, err = os.Create(mdlName + ".mtl")
	if err != nil {
		panic(err)
	}
	w = bufio.NewWriter(o)

	t, _ = template.New("").Parse(mtl)
	t.Execute(w, data)
	w.Flush()
	o.Close()
}

const obj = `o {{ .Name }}
mtllib {{ .Name }}.mtl
usemtl {{ .Name }}
{{range .V}}v {{ index . 0 }} {{ index . 1 }} {{ index . 2 }}
{{end}}vn -0.525731  0.000000  0.850651  
vn -0.442863  0.238856  0.864188  
vn -0.295242  0.000000  0.955423  
vn -0.309017  0.500000  0.809017  
vn -0.162460  0.262866  0.951056  
vn  0.000000  0.000000  1.000000  
vn  0.000000  0.850651  0.525731  
vn -0.147621  0.716567  0.681718  
vn  0.147621  0.716567  0.681718  
vn  0.000000  0.525731  0.850651  
vn  0.309017  0.500000  0.809017  
vn  0.525731  0.000000  0.850651  
vn  0.295242  0.000000  0.955423  
vn  0.442863  0.238856  0.864188  
vn  0.162460  0.262866  0.951056  
vn -0.681718  0.147621  0.716567  
vn -0.809017  0.309017  0.500000  
vn -0.587785  0.425325  0.688191  
vn -0.850651  0.525731  0.000000  
vn -0.864188  0.442863  0.238856  
vn -0.716567  0.681718  0.147621  
vn -0.688191  0.587785  0.425325  
vn -0.500000  0.809017  0.309017  
vn -0.238856  0.864188  0.442863  
vn -0.425325  0.688191  0.587785  
vn -0.716567  0.681718 -0.147621  
vn -0.500000  0.809017 -0.309017  
vn -0.525731  0.850651  0.000000  
vn  0.000000  0.850651 -0.525731  
vn -0.238856  0.864188 -0.442863  
vn  0.000000  0.955423 -0.295242  
vn -0.262866  0.951056 -0.162460  
vn  0.000000  1.000000  0.000000  
vn  0.000000  0.955423  0.295242  
vn -0.262866  0.951056  0.162460  
vn  0.238856  0.864188  0.442863  
vn  0.262866  0.951056  0.162460  
vn  0.500000  0.809017  0.309017  
vn  0.238856  0.864188 -0.442863  
vn  0.262866  0.951056 -0.162460  
vn  0.500000  0.809017 -0.309017  
vn  0.850651  0.525731  0.000000  
vn  0.716567  0.681718  0.147621  
vn  0.716567  0.681718 -0.147621  
vn  0.525731  0.850651  0.000000  
vn  0.425325  0.688191  0.587785  
vn  0.864188  0.442863  0.238856  
vn  0.688191  0.587785  0.425325  
vn  0.809017  0.309017  0.500000  
vn  0.681718  0.147621  0.716567  
vn  0.587785  0.425325  0.688191  
vn  0.955423  0.295242  0.000000  
vn  1.000000  0.000000  0.000000  
vn  0.951056  0.162460  0.262866  
vn  0.850651 -0.525731  0.000000  
vn  0.955423 -0.295242  0.000000  
vn  0.864188 -0.442863  0.238856  
vn  0.951056 -0.162460  0.262866  
vn  0.809017 -0.309017  0.500000  
vn  0.681718 -0.147621  0.716567  
vn  0.850651  0.000000  0.525731  
vn  0.864188  0.442863 -0.238856  
vn  0.809017  0.309017 -0.500000  
vn  0.951056  0.162460 -0.262866  
vn  0.525731  0.000000 -0.850651  
vn  0.681718  0.147621 -0.716567  
vn  0.681718 -0.147621 -0.716567  
vn  0.850651  0.000000 -0.525731  
vn  0.809017 -0.309017 -0.500000  
vn  0.864188 -0.442863 -0.238856  
vn  0.951056 -0.162460 -0.262866  
vn  0.147621  0.716567 -0.681718  
vn  0.309017  0.500000 -0.809017  
vn  0.425325  0.688191 -0.587785  
vn  0.442863  0.238856 -0.864188  
vn  0.587785  0.425325 -0.688191  
vn  0.688191  0.587785 -0.425325  
vn -0.147621  0.716567 -0.681718  
vn -0.309017  0.500000 -0.809017  
vn  0.000000  0.525731 -0.850651  
vn -0.525731  0.000000 -0.850651  
vn -0.442863  0.238856 -0.864188  
vn -0.295242  0.000000 -0.955423  
vn -0.162460  0.262866 -0.951056  
vn  0.000000  0.000000 -1.000000  
vn  0.295242  0.000000 -0.955423  
vn  0.162460  0.262866 -0.951056  
vn -0.442863 -0.238856 -0.864188  
vn -0.309017 -0.500000 -0.809017  
vn -0.162460 -0.262866 -0.951056  
vn  0.000000 -0.850651 -0.525731  
vn -0.147621 -0.716567 -0.681718  
vn  0.147621 -0.716567 -0.681718  
vn  0.000000 -0.525731 -0.850651  
vn  0.309017 -0.500000 -0.809017  
vn  0.442863 -0.238856 -0.864188  
vn  0.162460 -0.262866 -0.951056  
vn  0.238856 -0.864188 -0.442863  
vn  0.500000 -0.809017 -0.309017  
vn  0.425325 -0.688191 -0.587785  
vn  0.716567 -0.681718 -0.147621  
vn  0.688191 -0.587785 -0.425325  
vn  0.587785 -0.425325 -0.688191  
vn  0.000000 -0.955423 -0.295242  
vn  0.000000 -1.000000  0.000000  
vn  0.262866 -0.951056 -0.162460  
vn  0.000000 -0.850651  0.525731  
vn  0.000000 -0.955423  0.295242  
vn  0.238856 -0.864188  0.442863  
vn  0.262866 -0.951056  0.162460  
vn  0.500000 -0.809017  0.309017  
vn  0.716567 -0.681718  0.147621  
vn  0.525731 -0.850651  0.000000  
vn -0.238856 -0.864188 -0.442863  
vn -0.500000 -0.809017 -0.309017  
vn -0.262866 -0.951056 -0.162460  
vn -0.850651 -0.525731  0.000000  
vn -0.716567 -0.681718 -0.147621  
vn -0.716567 -0.681718  0.147621  
vn -0.525731 -0.850651  0.000000  
vn -0.500000 -0.809017  0.309017  
vn -0.238856 -0.864188  0.442863  
vn -0.262866 -0.951056  0.162460  
vn -0.864188 -0.442863  0.238856  
vn -0.809017 -0.309017  0.500000  
vn -0.688191 -0.587785  0.425325  
vn -0.681718 -0.147621  0.716567  
vn -0.442863 -0.238856  0.864188  
vn -0.587785 -0.425325  0.688191  
vn -0.309017 -0.500000  0.809017  
vn -0.147621 -0.716567  0.681718  
vn -0.425325 -0.688191  0.587785  
vn -0.162460 -0.262866  0.951056  
vn  0.442863 -0.238856  0.864188  
vn  0.162460 -0.262866  0.951056  
vn  0.309017 -0.500000  0.809017  
vn  0.147621 -0.716567  0.681718  
vn  0.000000 -0.525731  0.850651  
vn  0.425325 -0.688191  0.587785  
vn  0.587785 -0.425325  0.688191  
vn  0.688191 -0.587785  0.425325  
vn -0.955423  0.295242  0.000000  
vn -0.951056  0.162460  0.262866  
vn -1.000000  0.000000  0.000000  
vn -0.850651  0.000000  0.525731  
vn -0.955423 -0.295242  0.000000  
vn -0.951056 -0.162460  0.262866  
vn -0.864188  0.442863 -0.238856  
vn -0.951056  0.162460 -0.262866  
vn -0.809017  0.309017 -0.500000  
vn -0.864188 -0.442863 -0.238856  
vn -0.951056 -0.162460 -0.262866  
vn -0.809017 -0.309017 -0.500000  
vn -0.681718  0.147621 -0.716567  
vn -0.681718 -0.147621 -0.716567  
vn -0.850651  0.000000 -0.525731  
vn -0.688191  0.587785 -0.425325  
vn -0.587785  0.425325 -0.688191  
vn -0.425325  0.688191 -0.587785  
vn -0.425325 -0.688191 -0.587785  
vn -0.587785 -0.425325 -0.688191  
vn -0.688191 -0.587785 -0.425325 
{{range .VT}}vt {{ index . 0 }} {{ index . 1 }}
{{end}}{{range .F}}f {{ index . 0 }}/{{ index . 1 }}/{{ index . 2 }} {{ index . 3 }}/{{ index . 4 }}/{{ index . 5 }} {{ index . 6 }}/{{ index . 7 }}/{{ index . 8 }}
{{end}}`

const mtl = `newmtl {{ .Name }}
Ka 1.000000 1.000000 1.000000
Kd 1.000000 1.000000 1.000000
Ks 0.000000 0.000000 0.000000
Tr 1.000000
illum 1
Ns 0.000000
map_Kd {{ .Name }}.jpg
`
